sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java

```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
+     * the tests with a kinesis service emulator.
+     */
+    Write<T> withKinesisClientProvider(
+        AwsCredentialsProvider credentialsProvider, Region region, String endpoint) {
+      return withKinesisClientProvider(
+          new BasicKinesisProvider(credentialsProvider, region, endpoint));
+    }
+
+    /** Configure the AWS SDK retry behavior of the {@link KinesisAsyncClient}. */
+    public Write<T> withRetryConfiguration(RetryConfiguration retryConfiguration) {
+      return builder().retryConfiguration(retryConfiguration).build();
+    }
+
+    public Write<T> validated() {
+      checkArgument(!isEmpty(streamName()), "streamName cannot be empty");
+      checkArgument(partitioner() != null, "partitioner cannot be null");
+      checkArgument(serializer() != null, "serializer cannot be null");
+      checkArgument(kinesisClientProvider() != null, "kinesisClientProvider cannot be null");
+      checkArgument(concurrentRequests() > 0, "concurrentRequests must be > 0");
+      checkArgument(
+          batchMaxRecords() > 0 && batchMaxRecords() <= MAX_RECORDS_PER_REQUEST,
+          "batchMaxRecords must be in [1,%s]",
+          MAX_RECORDS_PER_REQUEST);
+      checkArgument(
+          batchMaxBytes() > 0 && batchMaxBytes() <= MAX_BYTES_PER_REQUEST,
+          "batchMaxBytes must be in [1,%s]",
+          MAX_BYTES_PER_REQUEST);
+      checkArgument(
+          !aggregatedRecords() || aggregatedRecordsMaxBytes() <= MAX_BYTES_PER_RECORD,
+          "aggregatedRecordsMaxBytes must be <= %s",
+          MAX_BYTES_PER_RECORD);
+      checkArgument(
+          !aggregatedRecords() || aggregatedRecordsMaxBytes() <= batchMaxBytes(),
+          "aggregatedRecordsMaxBytes must be <= batchMaxBytes (%s)",
+          batchMaxBytes());
+      return this;
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      validated();
+      if (aggregatedRecords()) {
+        input.apply(ParDo.of(new AggregatedWriterFn<T>(this)));
+      } else {
+        input.apply(ParDo.of(new WriterFn<T>(this)));
+      }
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriterFn<T> extends DoFn<T, Void> {
+      private static final int MIN_PARTIAL_RETRIES = 10; // Retries for partial success (throttling)
+      private static final ClientPool<
+              KinesisAsyncClientProvider, RetryConfiguration, KinesisAsyncClient>
+          clients = ClientPool.autoClosable((p, c) -> p.getKinesisAsyncClient(clientOverride(c)));
+
+      private static ClientOverrideConfiguration clientOverride(RetryConfiguration retryConfig) {
+        return (retryConfig == null) ? null : retryConfig.toClientOverride();
+      }
+
+      protected final Write<T> spec;
+
+      private transient AsyncPutRecordsHandler handler;
+      private transient List<PutRecordsRequestEntry> requestEntries;
+      private transient int requestBytes;
```

Comment:
why this field is `transient`?

Reply:
ping

Reply:
Both, `requestEntries` and `requestBytes` are state that's only relevant / valid while processing a bundle.